File: C++/source/teraterm/ttplug.c
#include "teraterm.h"
#include "tttypes.h"

// #include <windows.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
/* for _findXXXX() functions */
#include <io.h>
#include "ttwinman.h"
#include "ttplugin.h"
#include "ttplug.h"
#undef TTXOpenTCP
#undef TTXCloseTCP
#undef TTXGetUIHooks
#undef TTXGetSetupHooks

#define MAXNUMEXTENSIONS 16
static HANDLE LibHandle[MAXNUMEXTENSIONS];
static int NumExtensions = 0;
static TTXExports * * Extensions;

typedef struct _ExtensionList {
  TTXExports * exports;
  struct _ExtensionList * next;
} ExtensionList;

static int compareOrder(const void * e1, const void * e2) {
  TTXExports * * exports1 = (TTXExports * *)e1;
  TTXExports * * exports2 = (TTXExports * *)e2;

  return (*exports1)->loadOrder - (*exports2)->loadOrder;
}

static void loadExtension(ExtensionList * * extensions, char const * fileName) {
  char buf[1024];

  if (NumExtensions>=MAXNUMEXTENSIONS) return;
  LibHandle[NumExtensions] = LoadLibrary(fileName);
#ifdef TERATERM32
  if (LibHandle[NumExtensions] != NULL) {
    TTXBindProc bind = (TTXBindProc)GetProcAddress(LibHandle[NumExtensions], "_TTXBind@8");
    if (bind==NULL)
      bind = (TTXBindProc)GetProcAddress(LibHandle[NumExtensions], "TTXBind");
#else
  if (LibHandle[NumExtensions] >= HINSTANCE_ERROR) {
    TTXBindProc bind = (TTXBindProc)GetProcAddress(LibHandle[NumExtensions], "TTXBIND");
#endif
    if (bind != NULL) {
      ExtensionList * newExtension =
        (ExtensionList *)malloc(sizeof(ExtensionList));

      newExtension->exports = (TTXExports *)malloc(sizeof(TTXExports));
      memset(newExtension->exports, 0, sizeof(TTXExports));
      newExtension->exports->size = sizeof(TTXExports);
      if (bind(TTVERSION,(TTXExports FAR *)newExtension->exports)) {
        newExtension->next = *extensions;
        *extensions = newExtension;
        NumExtensions++;
        return;
      } else {
	free(newExtension->exports);
	free(newExtension);
      }
    }
    FreeLibrary(LibHandle[NumExtensions]);
  }

  _snprintf(buf, sizeof(buf), "Cannot load extension %s", fileName);
  buf[sizeof(buf) - 1] = 0;
  MessageBox(NULL, buf, "Teraterm Error", MB_OK | MB_ICONEXCLAMATION);
}

void PASCAL FAR TTXInit(PTTSet ts, PComVar cv) {
  ExtensionList * extensionList = NULL;
  int i;

  if (getenv("TERATERM_EXTENSIONS") != NULL) {
    char buf[1024];
    int index;
    struct _finddata_t searchData;
    long searchHandle;

    GetModuleFileName(hInst, buf, sizeof(buf));
    for (index = strlen(buf) - 1;
    index >= 0 && buf[index] != '\\' && buf[index] != ':' && buf[index] != '/';
      index--) {
    }
    index++;
    strncpy(buf + index, "TTX*.DLL", sizeof(buf) - index);

    searchHandle = _findfirst(buf,&searchData);
    if (searchHandle != -1L) {
      loadExtension(&extensionList, searchData.name);

      while (_findnext(searchHandle, &searchData)==0) {
	loadExtension(&extensionList, searchData.name);
      }
      _findclose(searchHandle);
    }

    if (NumExtensions==0) return;

    Extensions = (TTXExports * *)malloc(sizeof(TTXExports *)*NumExtensions);
    for (i = 0; i < NumExtensions; i++) {
      ExtensionList * old;

      Extensions[i] = extensionList->exports;
      old = extensionList;
      extensionList = extensionList->next;
      free(old);
    }

    qsort(Extensions, NumExtensions, sizeof(Extensions[0]), compareOrder);

    for (i = 0; i < NumExtensions; i++) {
      if (Extensions[i]->TTXInit != NULL) {
        Extensions[i]->TTXInit(ts, cv);
      }
    }
  }
}

void PASCAL FAR TTXInternalOpenTCP(TTXSockHooks FAR * hooks) {
  int i;

  for (i = 0; i < NumExtensions; i++) {
    if (Extensions[i]->TTXOpenTCP != NULL) {
      Extensions[i]->TTXOpenTCP(hooks);
    }
  }
}

void PASCAL FAR TTXInternalCloseTCP(TTXSockHooks FAR * hooks) {
  int i;

  for (i = NumExtensions - 1; i >= 0; i--) {
    if (Extensions[i]->TTXCloseTCP != NULL) {
      Extensions[i]->TTXCloseTCP(hooks);
    }
  }
}

void PASCAL FAR TTXInternalGetUIHooks(TTXUIHooks FAR * hooks) {
  int i;

  for (i = 0; i < NumExtensions; i++) {
    if (Extensions[i]->TTXGetUIHooks != NULL) {
      Extensions[i]->TTXGetUIHooks(hooks);
    }
  }
}

void PASCAL FAR TTXInternalGetSetupHooks(TTXSetupHooks FAR * hooks) {
  int i;

  for (i = NumExtensions - 1; i >= 0; i--) {
    if (Extensions[i]->TTXGetSetupHooks != NULL) {
      Extensions[i]->TTXGetSetupHooks(hooks);
    }
  }
}

void PASCAL FAR TTXSetWinSize(int rows, int cols) {
  int i;

  for (i = 0; i < NumExtensions; i++) {
    if (Extensions[i]->TTXSetWinSize != NULL) {
      Extensions[i]->TTXSetWinSize(rows, cols);
    }
  }
}

void PASCAL FAR TTXModifyMenu(HMENU menu) {
  int i;

  for (i = 0; i < NumExtensions; i++) {
    if (Extensions[i]->TTXModifyMenu != NULL) {
      Extensions[i]->TTXModifyMenu(menu);
    }
  }
}

void PASCAL FAR TTXModifyPopupMenu(HMENU menu) {
  int i;

  for (i = 0; i < NumExtensions; i++) {
    if (Extensions[i]->TTXModifyPopupMenu != NULL) {
      Extensions[i]->TTXModifyPopupMenu(menu);
    }
  }
}

BOOL PASCAL FAR TTXProcessCommand(HWND hWin, WORD cmd) {
  int i;

  for (i = NumExtensions - 1; i >= 0; i--) {
    if (Extensions[i]->TTXProcessCommand != NULL) {
      if (Extensions[i]->TTXProcessCommand(hWin,cmd)) {
        return TRUE;
      }
    }
  }

  return FALSE;
}

void PASCAL FAR TTXEnd(void) {
  int i;

  if (NumExtensions==0) return;

  for (i = NumExtensions - 1; i >= 0; i--) {
    if (Extensions[i]->TTXEnd != NULL) {
      Extensions[i]->TTXEnd();
    }
  }

  for (i=0; i<NumExtensions; i++)
    FreeLibrary(LibHandle[i]);

  free(Extensions);
  NumExtensions = 0;
}

void PASCAL FAR TTXSetCommandLine(PCHAR cmd, int cmdlen, PGetHNRec rec) {
  int i;

  for (i = 0; i < NumExtensions; i++) {
    if (Extensions[i]->TTXSetCommandLine != NULL) {
      Extensions[i]->TTXSetCommandLine(cmd, cmdlen, rec);
    }
  }
}

File: Pascal/ttplug.pas
{ Teraterm extension mechanism
   Robert O'Callahan (roc+tt@cs.cmu.edu)
   
   Teraterm by Takashi Teranishi (teranishi@rikaxp.riken.go.jp)
}
unit TTPlug;

interface

uses WinTypes, WinProcs, Strings, WinDos,
     Types, TTTypes, TTLib;

procedure TTXInit(ts: PTTSet; cv: PComVar);
procedure TTXOpenTCP;
procedure TTXCloseTCP;
procedure TTXGetUIHooks;
procedure TTXGetSetupHooks;
procedure TTXSetWinSize(rows, cols: integer);
procedure TTXModifyMenu(menu: HMENU);
procedure TTXModifyPopupMenu(menu: HMENU);
function  TTXProcessCommand(hWin: HWND; cmd: WORD): BOOL;
procedure TTXEnd;
procedure TTXSetCommandLine(cmd: PCHAR; cmdlen: integer; rec: PGetHNRec);

implementation

uses TTXTypes, TTWsk, TTSetup, TTDialog;
{$I teraterm.inc}

type
  PArrayTTXExports = ^ArrayTTXExports;
  ArrayTTXExports = array[0..255] of TTXExports; {array size is dummy}

  TTTXBindProc = function(Version: WORD; _exports: PTTXExports): BOOL;

const
  MAXNUMEXTENSIONS = 16;
var
  LibHandle: array[0..MAXNUMEXTENSIONS-1] of THandle;
  NumExtensions: integer;
  Extensions: PArrayTTXExports;

type
  PExtensionList = ^ExtensionList;
  ExtensionList = record
    _exports: PTTXExports;
    next: PExtensionList;
  end;

procedure SortExtensions;

  function Value(i: integer): integer;
  begin
    Value := Extensions^[i].loadOrder;
  end;

  procedure QSort(l,r: integer);
  var
    il,ir,v,t: integer;
    Tmp: TTXExports;
  begin
    il:=l;
    ir:=r;
    v := Value((l+r) div 2);
    repeat
      while Value(il)<v do inc(il);
      while v<Value(ir) do dec(ir);
      if il<=ir then
      begin
        Tmp := Extensions^[il];
        Extensions^[il] := Extensions^[ir];
        Extensions^[ir] := Tmp;
        inc(il);
        dec(ir);
      end;
    until il>ir;
    if l<ir then QSort(l,ir);
    if il<r then QSort(il,r);  
  end;

begin
  if NumExtensions<2 then exit;
  QSort(0,NumExtensions-1);
end;

procedure loadExtension(var extlist: PExtensionList; fileName: PCHAR);
var
  buf: array[0..1023] of char;
  bind: TTTXBindProc;
  newExtension: PExtensionList;
begin
  if NumExtensions>=MAXNUMEXTENSIONS then exit;
  LibHandle[NumExtensions] := LoadLibrary(fileName);

{$ifdef TERATERM32}
  if LibHandle[NumExtensions] <> nil then
  begin
    @bind := GetProcAddress(LibHandle[NumExtensions], '_TTXBind@4');
    if @bind=nil then
      @bind := GetProcAddress(LibHandle[NumExtensions],'TTXBind');
{$else}
  if LibHandle[NumExtensions] >=HINSTANCE_ERROR then
  begin
    @bind := GetProcAddress(LibHandle[NumExtensions], 'TTXBIND');
{$endif}
    if @bind <> nil then
    begin
      GetMem(newExtension,sizeof(ExtensionList));

      GetMem(newExtension^._exports,sizeof(TTXExports));
      fillchar(newExtension^._exports^,sizeof(TTXExports),0);
      newExtension^._exports^.size := sizeof(TTXExports);
      if bind(TTVERSION,newExtension^._exports) then
      begin
        newExtension^.next := extlist;
        extlist := newExtension;
        inc(NumExtensions);
        exit;
      end
      else begin
        FreeMem(newExtension^._exports,sizeof(TTXExports));
        FreeMem(newExtension,sizeof(ExtensionList));
      end;
    end;
    FreeLibrary(LibHandle[NumExtensions]);
  end;

  strcopy(buf,'Cannot load extension ');
  strcat(buf,filename);
  MessageBox(0, buf, 'Teraterm Error', MB_OK or MB_ICONEXCLAMATION);
end;

procedure TTXInit(ts: PTTSet; cv: PComVar);
var
  extList, old: PExtensionList;
  i: integer;
  buf: array[0..1023] of char;
  index: integer;
  searchData: TSearchRec;
begin
  extList := nil;

  if GetEnvVar('TERATERM_EXTENSIONS')<>nil then
  begin
    GetModuleFileName(hinstance, buf, sizeof(buf));
    GetFileNamePos(buf,index,i);
    buf[index] := #0;
    AppendSlash(buf);
    strcat(buf,'TTX*.DLL');
    FindFirst(buf,faArchive,searchData);
    while DosError=0 do
    begin
      loadExtension(extList,searchData.Name);
      FindNext(searchData);
    end;

    if NumExtensions=0 then exit;

    GetMem(Extensions,sizeof(TTXExports)*NumExtensions);
    for i := 0 to NumExtensions-1 do
    begin
      Move(extList^._exports^,Extensions^[i],sizeof(TTXExports));
      old := extList;
      extList := extList^.next;
      FreeMem(old,sizeof(ExtensionList));
    end;

    SortExtensions;

    for i := 0 to NumExtensions-1 do
    begin
      if @Extensions^[i].TTXInit <> nil then
        Extensions^[i].TTXInit(ts, cv);
    end;
  end;
end;

procedure TTXInternalOpenTCP(hooks: PTTXSockHooks);
var
  i: integer;
begin
  for i := 0 to NumExtensions-1 do
  begin
    if @Extensions^[i].TTXOpenTCP <> nil then
      Extensions^[i].TTXOpenTCP(hooks);
  end;
end;

procedure TTXOpenTCP;
var
  h: TTXSockHooks;
begin
  @h.closesocket := @closesocket;
  @h.connect := @connect;
  @h.htonl := @htonl;
  @h.htons := @htons;
  @h.inet_addr := @inet_addr;
  @h.ioctlsocket := @ioctlsocket;
  @h.recv := @recv;
  @h.select := @select;
  @h.send := @send;
  @h.setsockopt := @setsockopt;
  @h._socket := @_socket;
  @h.WSAAsyncSelect := @WSAAsyncSelect;
  @h.WSAAsyncGetHostByName := @WSAAsyncGetHostByName;
  @h.WSACancelAsyncRequest := @WSACancelAsyncRequest;
  @h.WSAGetLastError := @WSAGetLastError;
  TTXInternalOpenTCP(@h);
end;

procedure TTXInternalCloseTCP(hooks: PTTXSockHooks);
var
  i: integer;
begin
  for i := NumExtensions-1 downto 0 do
  begin
    if @Extensions^[i].TTXCloseTCP <> nil then
      Extensions^[i].TTXCloseTCP(hooks);
  end;
end;

procedure TTXCloseTCP;
var
  h: TTXSockHooks;
begin
  @h.closesocket := @closesocket;
  @h.connect := @connect;
  @h.htonl := @htonl;
  @h.htons := @htons;
  @h.inet_addr := @inet_addr;
  @h.ioctlsocket := @ioctlsocket;
  @h.recv := @recv;
  @h.select := @select;
  @h.send := @send;
  @h.setsockopt := @setsockopt;
  @h._socket := @_socket;
  @h.WSAAsyncSelect := @WSAAsyncSelect;
  @h.WSAAsyncGetHostByName := @WSAAsyncGetHostByName;
  @h.WSACancelAsyncRequest := @WSACancelAsyncRequest;
  @h.WSAGetLastError := @WSAGetLastError;
  TTXInternalCloseTCP(@h);
end;

procedure TTXInternalGetUIHooks(hooks: PTTXUIHooks);
var
  i: integer;
begin
  for i := 0 to NumExtensions-1 do
  begin
    if @Extensions^[i].TTXGetUIHooks <> nil then
      Extensions^[i].TTXGetUIHooks(hooks);
  end;
end;

procedure TTXGetUIHooks;
var
  h: TTXUIHooks;
begin
  @h.SetupTerminal := @SetupTerminal;
  @h.SetupWin := @SetupWin;
  @h.SetupKeyboard := @SetupKeyboard;
  @h.SetupSerialPort := @SetupSerialPort;
  @h.SetupTCPIP := @SetupTCPIP;
  @h.GetHostName := @GetHostName;
  @h.ChangeDirectory := @ChangeDirectory;
  @h.AboutDialog := @AboutDialog;
  @h.ChooseFontDlg := @ChooseFontDlg;
  @h.SetupGeneral := @SetupGeneral;
  @h.WindowWindow := @WindowWindow;
  TTXInternalGetUIHooks(@h);
end;

procedure TTXInternalGetSetupHooks(hooks: PTTXSetupHooks);
var
  i: integer;
begin
  for i := NumExtensions-1 downto 0 do
  begin
    if @Extensions^[i].TTXGetSetupHooks <> nil then
      Extensions^[i].TTXGetSetupHooks(hooks);
  end;
end;

procedure TTXGetSetupHooks;
var
  h: TTXSetupHooks;
begin
  @h.ReadIniFile := @ReadIniFile;
  @h.WriteIniFile := @WriteIniFile;
  @h.ReadKeyboardCnf := @ReadKeyboardCnf;
  @h.CopyHostList := @CopyHostList;
  @h.AddHostToList := @AddHostToList;
  @h.ParseParam := @ParseParam;
  TTXInternalGetSetupHooks(@h);
end;

procedure TTXSetWinSize(rows, cols: integer);
var
  i: integer;
begin
  for i := 0 to NumExtensions-1 do
  begin
    if @Extensions^[i].TTXSetWinSize <> nil then
      Extensions^[i].TTXSetWinSize(rows, cols);
  end;
end;

procedure TTXModifyMenu(menu: HMENU);
var
  i: integer;
begin
  for i := 0 to NumExtensions-1 do
  begin
    if @Extensions^[i].TTXModifyMenu <> nil then
      Extensions^[i].TTXModifyMenu(menu);
  end;
end;

procedure TTXModifyPopupMenu(menu: HMENU);
var
  i: integer;
begin
  for i := 0 to NumExtensions-1 do
  begin
    if @Extensions^[i].TTXModifyPopupMenu <> nil then
      Extensions^[i].TTXModifyPopupMenu(menu);
  end;
end;

function TTXProcessCommand(hWin: HWND; cmd: WORD): BOOL;
var
  i: integer;
begin
  for i := NumExtensions-1 downto 0 do
  begin
    if @Extensions^[i].TTXProcessCommand <> nil then
    begin
      if Extensions^[i].TTXProcessCommand(hWin,cmd)<>0 then
      begin
        TTXProcessCommand := TRUE;
        exit;
      end;
    end;
  end;
  TTXProcessCommand := FALSE;
end;

procedure TTXEnd;
var
  i: integer;
begin
  if NumExtensions=0 then exit;

  for i := NumExtensions-1 downto 0 do
  begin
    if @Extensions^[i].TTXEnd <> nil then
      Extensions^[i].TTXEnd;
  end;
  for i := 0 to NumExtensions-1 do
    FreeLibrary(LibHandle[i]);

  FreeMem(Extensions,sizeof(TTXExports)*NumExtensions);
  NumExtensions := 0;
end;

procedure TTXSetCommandLine(cmd: PCHAR; cmdlen: integer; rec: PGetHNRec);
var
  i: integer;
begin
  for i := 0 to NumExtensions-1 do
  begin
    if @Extensions^[i].TTXSetCommandLine <> nil then
      Extensions^[i].TTXSetCommandLine(cmd,cmdlen,rec);
  end;
end;

begin
  NumExtensions := 0;
end.
